### PR TITLE
Add initial guess as centroid

### PR DESCRIPTION
**Pull request by gandroz**
According to the issue #58 , here a proposal to improve clustering (only for the KShape method for now) by letting the user choose an initial guess as centroids. This guess is a numpy array of int which are the indices of the samples to be used as centroids instead of a random vector.

### CODE REVIEW

**Comment by GillesVandewiele** [comment]
Hello, great work! Small suggestion: wouldn't it be better to let the user provide initial centroids, instead of indices? That way, you are not bound to have the initial centroids set to time series that are WITHIN your dataset. The suggested approach works well for your specific use case, where you know a sample of each cluster, but imagine if you would know more than 1 sample, then you could calculate a centroid of these samples and provide it to the algorithm.

**Comment by gandroz** [comment]
You're right, it is just a "quick win" I implemented on my side. Your suggestion is interesting and that was my first though, but I did not want to deal with time series of different lengths for example, and normalization would have been a bit more difficult as we have to normalize the initial guess the same manner the X_train is normalized, but not too difficult.

I could check that on my side if you want ?

**Comment by GillesVandewiele** [comment]
Does the algorithm currently support time series of variable length? I do not really know the KShape algorithm that well... Does not seem to give an error when I try it in my Python shell with a simple example.

Indeed, I did not think of the normalization, but it should be easy to fix if we decouple the line on https://github.com/rtavenar/tslearn/pull/59/commits/f4db8cdcc33256cf2c5d676e36adde97861dd66d#diff-79bececde0dc56f7672e4cedbaa36832R750 such that we have the `TimeSeriesScalerMeanVariance` object, then we can access the required data to normalize the provided initial centroid as wel.

**Comment by GillesVandewiele** [comment]
And on another note: maybe a few asserts should be added, such that a fitting exception can be raised? One that comes to mind straight away is that the number of `initial centroids` must be equal to `n_clusters`. But maybe other assertions concerning the length of these centroids can be made as well?

**Comment by gandroz** [comment]
Normalization is not a big deal indeed. I'll fix that. And yes, KShape support time series of variable length as it uses the method `to_time_serries_dataset` and not `to_time_series`.
There is already an assertion on the length of the initial guess at [line 753 ](https://github.com/rtavenar/tslearn/commit/f4db8cdcc33256cf2c5d676e36adde97861dd66d#diff-79bececde0dc56f7672e4cedbaa36832R754)

And a question for you, the tests failed on Travis. Is there a way I can run them before pushing ?

**Comment by GillesVandewiele** [comment]
My apologies, I looked over that! It seems like the doctest of clustering.py is failing...

You can check the logs (https://travis-ci.org/rtavenar/tslearn)[here] and to run the doctest, just run `python3 -m doctest -v clustering.py` in the `tslearn` directory.

**Comment by gandroz** [comment]
Hi,
I've just pushed some changes as you suggested. You can now use an initial centroids array instead of an array of indices. I tested it on my case and it works as expected. 
I also modified the preprocessing code by vectorizing it more so that it is quickier now.

**Comment by GillesVandewiele** [comment]
Great work mate! Seems like the doctests are passing as well. Looks like a great addition to me!

**Comment by gandroz** [comment]
you welcome, it is really a nice package btw !

**Comment by GillesVandewiele** [comment]
All props to @rtavenar. I haven't contributed yet to this repository. But I completely agree, this package is great...

Hmmm Travis seems to have failed, while it was successful for python 2.7 :thinking: Do the doctests work locally for you?

**Comment by gandroz** [comment]
yes it does, but I use p27. Seem like it is a build issue 

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

**Comment by GillesVandewiele** [comment]
Ye but it's strange that it takes longer than 10 minutes... Other builds of python 3 mostly take around 5 minutes or something... Do you have python3 installed? Could you run a doctest with python3 possibly?

On the other hand, the build of python 3.5 just went terrible, it couldn't even install python 3.5

**Comment by gandroz** [comment]
The issue is really at the build time of the docker image. I could try the doctest with python 3.5 but the test does not event go that far

**Comment by GillesVandewiele** [comment]
Ye exactly, something really weird. I'm kinda clueless as well :/. In the other builds, it did seem to be executing the nosetests though.

```
$ nosetests $KERAS_IGNORE --with-doctest --with-coverage --cover-erase --cover-package=tslearn
Using TensorFlow backend.
...........
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

I think those dots correspond to successful tests.

**Comment by gandroz** [comment]
The tests passed for python 3.5, but there is a strange behaviour. Two tests are said `fail` but the result is the same as expected:

>  Expected:
    array([[[ 1. ],
            [ 1.5],
            [ 2. ]]])
Got:
    array([[[1. ],
            [1.5],
            [2. ]]])

The difference is just a white space before numbers, not the values....

File: tslearn/preprocessing.py [hunk 1]
```diff
@@ -131,10 +130,12 @@ class TimeSeriesScalerMeanVariance(TransformerMixin):
     def __init__(self, mu=0., std=1.):
         self.mu_ = mu
         self.std_ = std
+        self.global_mean = None
+        self.global_std = None
 
-    def fit_transform(self, X, **kwargs):
+    def fit_transform(self, X, global_stats=False, **kwargs):
```
Comment (rtavenar) [inline tslearn/preprocessing.py]: Hum, `global_stats` argument is never used afterwards, is it?

If not, you should remove it, I guess.

File: tslearn/clustering.py [hunk 2]
```diff
@@ -731,20 +734,27 @@ def _fit_one_init(self, X, rs):
...
         ----------
         X : array-like of shape=(n_ts, sz, d)
             Time series dataset.
+        initial_centroids: array-like of shape=(n_clusters, sz, d) of initial centroids
+            Time series dataset.
         """
 
         X_ = to_time_series_dataset(X)
         self._norms = numpy.linalg.norm(X_, axis=(1, 2))
-        X_ = TimeSeriesScalerMeanVariance(mu=0., std=1.).fit_transform(X_)
```
Comment (rtavenar) [inline tslearn/clustering.py]: Why is this line removed?

File: tslearn/clustering.py [hunk 3]
```diff
@@ -704,10 +704,13 @@ def _assign(self, X):
         _check_no_empty_cluster(self.labels_, self.n_clusters)
         self.inertia_ = _compute_inertia(dists, self.labels_)
 
-    def _fit_one_init(self, X, rs):
+    def _fit_one_init(self, X, initial_centroids, rs):
```
Comment (rtavenar) [inline tslearn/clustering.py]: Would it be possible to rename this argument as `init` and more or less match sklearn usage (cf. https://github.com/scikit-learn/scikit-learn/blob/f0ab589f/sklearn/cluster/k_means_.py#L707)?

File: tslearn/preprocessing.py [hunk 4]
```diff
@@ -146,11 +147,10 @@ def fit_transform(self, X, **kwargs):
...
         """
         X_ = to_time_series_dataset(X)
-        for i in range(X_.shape[0]):
-            for d in range(X_.shape[2]):
-                cur_mean = X_[i, :, d].mean()
-                cur_std = X_[i, :, d].std()
-                if cur_std == 0.:
-                    cur_std = 1.
-                X_[i, :, d] = (X_[i, :, d] - cur_mean) * self.std_ / cur_std + self.mu_
+        mean_t = numpy.mean(X_, axis=1)[:, numpy.newaxis, :]
```
Comment (GillesVandewiele) [inline tslearn/preprocessing.py]: I tried converting the code to numpy in the past and discovered that it actually ran slower than a Pure-Python loop. Maybe this should be tested more extensively, and if it indeed runs slower, convert back to the old code

Comment (rtavenar) [reply tslearn/preprocessing.py]: @GillesVandewiele : is it still slower for datasets with many time series? Because it definitely looks like a better solution than the original one in theory :)

File: tslearn/clustering.py (hunk 2 above)
Comment (GillesVandewiele) [reply tslearn/clustering.py]: Seems like the code has changed to 

`self._norms = numpy.linalg.norm(X_, axis=(1, 2))` and `self._norms_centroids = numpy.linalg.norm(self.cluster_centers_, axis=(1, 2))`. I would also use the Normalizers from tslearn instead of using the numpy norm function...

File: tslearn/preprocessing.py (hunk 1 above)
Comment (GillesVandewiele) [reply tslearn/preprocessing.py]: i think we actually made a reasoning mistake and thought the normalization happened dataset-wise, I think the global_stats and `self.global_mean` + `self.global_std` stem from that. Of course, this becomes unnecessary since the normalization happens timeseries-wise. We just have to use the `transform` method of the SAME normalizer on both the provided X input array as the centroids. 

Important to note in the documentation is that the centroids may not be normalized by the user himself.

File: tslearn/preprocessing.py (hunk 4 above)
Comment (GillesVandewiele) [reply tslearn/preprocessing.py]: I mentioned it in Issue #52 right before getting closed, so maybe you looked over it. My code was not exactly the same as this:

```
X_ = to_time_series_dataset(X)
means, stds = np.mean(X_, axis=1), np.std(X_, axis=1)
stds[stds == 0] = 1
return (X_ - means) * self.std_ / stds + self.mu_
```

This seemed to run slower when I did some simple tests, but maybe should be double checked. It is definitely cleaner code!

Comment (GillesVandewiele) [reply tslearn/preprocessing.py]: This conversion to numpy can btw also easily be done for the MinMaxScaler!

Comment (rtavenar) [reply tslearn/preprocessing.py]: Yep, but in Issue #52, you tested on a single long time series, which did not benefited from removing the loop. So hopefully with many time series that would be better!

Test:
```
time_series = numpy.random.randn(100000, 100, 10)
t0 = time.time()
TimeSeriesScalerMinMax().fit_transform(time_series)
t1 = time.time()
TimeSeriesScalerMinMaxTuned().fit_transform(time_series)
t2 = time.time()
```

With `TimeSeriesScalerMinMaxTuned` being the proposed implementation.

Gives:
```
12.474971771240234 5.250811815261841
```

So, good idea :)

**Comment by rtavenar** [comment]
The good news is: tests seem to have failed because of an issue on Travis CI, now they pass for all configs!

File: tslearn/preprocessing.py (hunk 4 above)
Comment (GillesVandewiele) [reply tslearn/preprocessing.py]: You are right! Mistake from my side! I benched it myself now as well :)  

numpy_times

python_times

**Comment by gandroz** [comment]
Hi sorry for the delay (i'm in vacation..). The numpy and vectorized
approach is definitely more efficient than the loop approach. It ran 10
times quickier in my tests so I think it worses the change.

Glad to ear the failed tests were due to Travis error.


Le ven. 17 août 2018 09:42, Gilles Vandewiele <notifications@github.com> a
écrit :

**Comment by gandroz** [comment]
Before adding the initial guess to the `TimeSeriesKMeans`, we tweaked the `KShape` method to shift back the centroid after computation of the eigen vectors to avoid cumulative shift fo the centroid and allow further optimization of the centroids. The visual result is that all centroids are better aligned to the samples. The performance result I suspect (not fully verified) is that because the centroids are back shifted, the cross-correlation is performed on a wider vector so that the NCC measure might be more representative.

**Comment by rtavenar** [comment]
I just merged it, but did not take the trick for kSHAPE into account: for me, it's unrelated to initial centroids, so I prefer to keep things separate.
